File: service/api/documents.py
"""The document resource: upload, browse, inspect, re-run, delete.

Serialisation is hand-written ``_row``/``_detail`` functions rather than
Pydantic response models — the reference project's convention, and it keeps the
wire format visible in one place instead of spread across model definitions.
Pydantic is still used for request bodies, where validation earns its keep.
"""
from __future__ import annotations

import logging
from typing import Any

from fastapi import (APIRouter, Depends, File, HTTPException, Query, Response,
                     UploadFile, status)
from fastapi.responses import FileResponse

from service import worker
from service.api.deps import require_api_or_session, require_session
from service.core.config import get_settings
from service.core.database import DbSession, get_db
from service.core.models import VALID_STATUSES, iso
from service.repositories import artifacts
from service.repositories import documents as repo

log = logging.getLogger(__name__)
router = APIRouter(prefix="/documents", tags=["documents"])

MAX_FILENAME_LEN = 200


def _safe_filename(raw: str | None) -> str:
    """Keep a display name only — it never touches the filesystem.

    Stored artifacts always use a fixed name (``original.jpg``), so even a
    hostile filename cannot escape the document directory. This is purely so
    the UI shows something sensible and bounded.
    """
    name = (raw or "upload").replace("\\", "/").split("/")[-1].strip()
    name = "".join(ch for ch in name if ch.isprintable() and ch not in '<>:"|?*')
    return (name or "upload")[:MAX_FILENAME_LEN]


def _row(record) -> dict[str, Any]:
    """One line of the document log."""
    return {
        "id": record.id,
        "filename": record.filename,
        "size_bytes": record.size_bytes,
        "status": record.status,
        "doc_type": record.doc_type,
        "doc_type_base": (record.doc_type or "").rsplit("_", 1)[0] or None
        if record.doc_type else None,
        "doc_type_era": (record.doc_type or "").rsplit("_", 1)[1]
        if record.doc_type and "_" in record.doc_type else None,
        "recognised": record.recognised,
        "doc_conf": record.doc_conf,
        "quality": record.quality or {},
        "field_count": record.field_count,
        "device": record.device,
        "processing_ms": record.processing_ms,
        "error": record.error,
        "error_code": record.error_code,
        "retry_count": record.retry_count,
        "has_canvas": record.has_canvas,
        "created_at": iso(record.created_at),
        "started_at": iso(record.started_at),
        "finished_at": iso(record.finished_at),
    }


def _detail(record) -> dict[str, Any]:
    """The row plus the full recognition view model, flattened into it.

    The stored ``result`` already has the client-facing shape (boxes, fields,
    canvas dimensions, coordinate-space notes) — see ``service/ml/transform.py``.
    """
    payload = _row(record)
    result = record.result or {}
    payload.update({
        "canvas": {
            **(result.get("canvas") or {}),
            "url": f"/api/v1/documents/{record.id}/image/canvas",
        },
        "original": {
            "url": f"/api/v1/documents/{record.id}/image/original",
            "width": record.original_w,
            "height": record.original_h,
            "content_type": record.content_type,
        },
        "coord_space": result.get("coord_space"),
        "coord_space_note": result.get("coord_space_note"),
        "boxes": result.get("boxes") or [],
        "fields": result.get("fields") or [],
        "ocr": result.get("ocr") or {},
        "quality": result.get("quality") or {},
        "timings": result.get("timings") or {},
        "address": result.get("address"),
    })
    return payload


@router.post("", status_code=status.HTTP_202_ACCEPTED)
async def upload_document(
    file: UploadFile = File(...),
    db: DbSession = Depends(get_db),
    _identity=Depends(require_api_or_session),
) -> dict[str, Any]:
    """Accept one image and queue it.

    Everything that can be checked cheaply is checked here, so a bad upload
    fails immediately with an actionable message instead of becoming a
    mysterious failed job a minute later.
    """
    settings = get_settings()
    limit = settings.max_upload_mb * 1024 * 1024

    data = await file.read(limit + 1)
    if len(data) > limit:
        raise HTTPException(status.HTTP_413_REQUEST_ENTITY_TOO_LARGE,
                            detail=f"File exceeds the {settings.max_upload_mb} MB limit")
    if not data:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, detail="Empty upload")

    if artifacts.is_pdf(data):
        # Called out separately because people will try it, and "unsupported
        # image type" would not tell them what to do about it.
        raise HTTPException(
            status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
            detail="PDF is not supported — upload a JPEG, PNG, WEBP, BMP or TIFF image")

    sniffed = artifacts.sniff_image(data)
    if sniffed is None:
        # Sniffed from magic bytes, not the client's Content-Type, which is
        # attacker-controlled and wrong often enough to be useless.
        raise HTTPException(status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
                            detail="Unsupported file type — expected an image")
    ext, media_type = sniffed

    dimensions = artifacts.decode_dimensions(data)
    if dimensions is None:
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_ENTITY,
                            detail="The image could not be decoded — it may be corrupt")

    filename = _safe_filename(file.filename)

    # Bytes first, row second. The record is what makes the document visible to
    # the worker, so writing it before the file leaves a window in which the
    # drain loop can claim a document whose original does not exist yet and mark
    # a perfectly good upload as failed. See repo.reserve_id.
    doc_id = repo.reserve_id(db)
    artifacts.save_original(db, doc_id, data, ext)
    record = repo.create(db, doc_id=doc_id, filename=filename,
                         content_type=media_type,
                         size_bytes=len(data), original_ext=ext,
                         original_w=dimensions[0], original_h=dimensions[1],
                         search_text=filename.lower())

    worker.notify_new_work()
    log.info("[API] queued doc=%s (%s, %d bytes)", record.id, filename, len(data))

    row = _row(record)
    row["queue_position"] = repo.queue_position(db, record.id)
    return row


@router.get("")
def list_documents(
    status_filter: str | None = Query(None, alias="status"),
    doc_type: str | None = Query(None),
    search: str | None = Query(None),
    date_from: str | None = Query(None),
    date_to: str | None = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    sort_by: str = Query("created_at"),
    sort_dir: str = Query("desc"),
    db: DbSession = Depends(get_db),
    _identity=Depends(require_api_or_session),
) -> dict[str, Any]:
    if status_filter and status_filter not in VALID_STATUSES:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, detail="Invalid status")
    if sort_dir not in ("asc", "desc"):
        sort_dir = "desc"

    rows, total = repo.get_all(db, status=status_filter, doc_type=doc_type,
                               search=search, date_from=date_from, date_to=date_to,
                               page=page, page_size=page_size,
                               sort_by=sort_by, sort_dir=sort_dir)
    return {
        "items": [_row(r) for r in rows],
        "total": total, "page": page, "page_size": page_size,
        "stats": repo.stats(db),
    }


@router.get("/{doc_id}")
def get_document(doc_id: int, db: DbSession = Depends(get_db),
                 _identity=Depends(require_api_or_session)) -> dict[str, Any]:
    record = repo.get_by_id(db, doc_id)
    if record is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="Document not found")
    return _detail(record)


@router.get("/{doc_id}/progress")
def get_progress(doc_id: int, db: DbSession = Depends(get_db),
                 _identity=Depends(require_api_or_session)) -> dict[str, Any] | None:
    """Live progress, queue position, or a terminal state.

    Returns ``200`` with a JSON ``null`` body when there is nothing to report —
    not 404, which would make the polling client raise an error toast every two
    seconds.
    """
    record = repo.get_by_id(db, doc_id)
    if record is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="Document not found")

    live = worker.get_document_progress(doc_id)
    if live is not None:
        return live

    if record.status == "queued":
        position = repo.queue_position(db, doc_id) or 0
        return {
            "step": "queued",
            "label": f"Queued (#{position + 1})",
            "pct": 0,
            "eta_sec": round(position * worker.average_duration_sec(), 1),
            "queue_position": position,
        }
    if record.status in ("done", "failed"):
        return {"step": record.status, "label": record.status.capitalize(),
                "pct": 100 if record.status == "done" else 0,
                "eta_sec": None, "queue_position": None}
    return None


@router.get("/{doc_id}/image/{kind}")
def get_image(doc_id: int, kind: str, db: DbSession = Depends(get_db),
              _identity=Depends(require_api_or_session)) -> Response:
    if kind not in ("original", "canvas", "thumb"):
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="Unknown image kind")
    found = artifacts.open_artifact(db, doc_id, kind)
    if found is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="Image not available")
    path, media_type = found
    # `no-cache` means "revalidate", not "do not store": FileResponse still sends
    # an ETag and Last-Modified, so a repeat request costs a 304 with no body.
    # `max-age=3600` was wrong here — Reprocess overwrites canvas.png and thumb.jpg
    # at the *same* URL, so the browser kept showing the previous recognition's
    # image for up to an hour while the field table beside it was already new.
    return FileResponse(path, media_type=media_type,
                        headers={"Cache-Control": "private, no-cache"})


@router.post("/{doc_id}/reprocess")
def reprocess_document(doc_id: int, db: DbSession = Depends(get_db),
                       _identity=Depends(require_api_or_session)) -> dict[str, Any]:
    record = repo.get_by_id(db, doc_id)
    if record is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="Document not found")
    if record.status in ("queued", "processing"):
        raise HTTPException(status.HTTP_409_CONFLICT,
                            detail=f"Document is already {record.status}")
    record = repo.requeue(db, record)
    worker.notify_new_work()
    return _row(record)


@router.delete("/{doc_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_document(doc_id: int, db: DbSession = Depends(get_db),
                    _identity=Depends(require_api_or_session)) -> Response:
    record = repo.get_by_id(db, doc_id)
    if record is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="Document not found")
    repo.delete(db, record)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/purge", status_code=status.HTTP_200_OK)
def purge_documents(db: DbSession = Depends(get_db),
                    _user=Depends(require_session)) -> dict[str, int]:
    """Clear the scratch store. Session-only — not something an integration does."""
    removed = 0
    for record in db.all_records():
        if record.status != "processing":  # leave an in-flight job alone
            repo.delete(db, record)
            removed += 1
    log.info("[API] purged %d document(s)", removed)
    return {"deleted": removed}

File: ports/go/internal/svc/api/documents.go
package api

import (
	"errors"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"strings"
	"unicode"

	"github.com/protei300/RussianDocsOCR/ports/go/internal/svc/errs"
	"github.com/protei300/RussianDocsOCR/ports/go/internal/svc/model"
	"github.com/protei300/RussianDocsOCR/ports/go/internal/svc/repo"
	"github.com/protei300/RussianDocsOCR/ports/go/internal/svc/store"
)

// The document resource: upload, browse, inspect, re-run, delete.
//
// Serialisation is hand-written `row`/`detail` functions rather than derived from the record
// struct, which is the reference's convention and keeps the wire format visible in ONE place
// instead of spread across tags on a type that also has to satisfy the store.

const maxFilenameLen = 200

// safeFilename keeps a DISPLAY NAME only — it never touches the filesystem.
//
// Stored artifacts always use a fixed name, so even a hostile filename cannot escape the
// document directory. This is purely so the UI shows something sensible and bounded; it is
// NOT the path-traversal defence, and treating it as one would be a mistake, because the real
// defence is that the name is never used as a path at all.
func safeFilename(raw string) string {
	name := raw
	if i := strings.LastIndexAny(strings.ReplaceAll(name, "\\", "/"), "/"); i >= 0 {
		name = strings.ReplaceAll(name, "\\", "/")[i+1:]
	}
	name = strings.TrimSpace(name)

	var b strings.Builder
	for _, r := range name {
		if !unicode.IsPrint(r) || strings.ContainsRune(`<>:"|?*`, r) {
			continue
		}
		b.WriteRune(r)
	}
	out := b.String()
	if out == "" {
		out = "upload"
	}
	// Truncated by RUNES, not bytes: a Cyrillic filename cut mid-rune renders as a
	// replacement character, and these names are routinely Cyrillic here.
	runes := []rune(out)
	if len(runes) > maxFilenameLen {
		runes = runes[:maxFilenameLen]
	}
	return string(runes)
}

// row is one line of the document log.
func row(rec *model.Document) map[string]any {
	var base, era *string
	if rec.DocType != nil {
		b, e := splitDocType(*rec.DocType)
		if b != "" {
			base = &b
		}
		if e != "" {
			era = &e
		}
	}
	quality := rec.Quality
	if quality == nil {
		quality = map[string]any{}
	}
	return map[string]any{
		"id":            rec.ID,
		"filename":      rec.Filename,
		"size_bytes":    rec.SizeBytes,
		"status":        rec.Status,
		"doc_type":      rec.DocType,
		"doc_type_base": base,
		"doc_type_era":  era,
		"recognised":    rec.Recognised,
		"doc_conf":      rec.DocConf,
		"quality":       quality,
		"field_count":   rec.FieldCount,
		"device":        rec.Device,
		"processing_ms": rec.ProcessingMs,
		"error":         rec.Error,
		"error_code":    rec.ErrorCode,
		"retry_count":   rec.RetryCount,
		"has_canvas":    rec.HasCanvas,
		"created_at":    rec.CreatedAt,
		"started_at":    rec.StartedAt,
		"finished_at":   rec.FinishedAt,
	}
}

func splitDocType(label string) (base, era string) {
	if i := strings.LastIndex(label, "_"); i >= 0 {
		return label[:i], label[i+1:]
	}
	return label, ""
}

// detail is the row plus the stored view model flattened into it.
//
// The stored result already has the client-facing shape — boxes, fields, canvas dimensions,
// coordinate-space notes — so this adds URLs and the original's dimensions and otherwise
// passes it through. Re-deriving any of it here would create a second definition of the wire
// format.
func detail(rec *model.Document) map[string]any {
	payload := row(rec)
	result := rec.Result
	if result == nil {
		result = map[string]any{}
	}

	canvas := map[string]any{}
	if c, ok := result["canvas"].(map[string]any); ok {
		for k, v := range c {
			canvas[k] = v
		}
	}
	canvas["url"] = fmt.Sprintf("/api/v1/documents/%d/image/canvas", rec.ID)

	payload["canvas"] = canvas
	payload["original"] = map[string]any{
		"url":          fmt.Sprintf("/api/v1/documents/%d/image/original", rec.ID),
		"width":        rec.OriginalW,
		"height":       rec.OriginalH,
		"content_type": rec.ContentType,
	}
	payload["coord_space"] = result["coord_space"]
	payload["coord_space_note"] = result["coord_space_note"]
	payload["boxes"] = orEmptyList(result["boxes"])
	payload["fields"] = orEmptyList(result["fields"])
	payload["ocr"] = orEmptyMap(result["ocr"])
	payload["quality"] = orEmptyMap(result["quality"])
	payload["timings"] = orEmptyMap(result["timings"])
	payload["address"] = result["address"]
	return payload
}

// orEmptyList and orEmptyMap keep a missing key from becoming a JSON null where the client
// expects a container. The SPA iterates `boxes` and `fields` unconditionally, so a null there
// is a runtime error in the browser rather than an empty table.
func orEmptyList(v any) any {
	if list, ok := v.([]any); ok && list != nil {
		return list
	}
	return []any{}
}

func orEmptyMap(v any) any {
	if m, ok := v.(map[string]any); ok && m != nil {
		return m
	}
	return map[string]any{}
}

// handleUpload accepts one image and queues it. 202 with the FULL LIST ROW, so the SPA can
// insert the row without a second request.
//
// Everything cheap is checked HERE, so a bad upload fails immediately with an actionable
// message instead of becoming a mysterious failed job a minute later.
func (s *Server) handleUpload(w http.ResponseWriter, r *http.Request, _ *Identity) {
	limit := s.cfg.MaxUploadBytes()
	// The body is capped BEFORE reading, so an oversized upload cannot exhaust memory
	// while being measured. Reading limit+1 is what distinguishes "at the limit" from
	// "over it".
	r.Body = http.MaxBytesReader(w, r.Body, limit+1)

	if err := r.ParseMultipartForm(8 << 20); err != nil {
		// **An oversized upload fails HERE, not at the size check below.** MaxBytesReader
		// aborts the read as soon as the cap is passed, so ParseMultipartForm is what returns
		// the error — and reporting it as a malformed request would tell the user to fix their
		// client when the actual problem is a 40 MB file. The size check further down only
		// catches a body that fit within the cap.
		var tooLarge *http.MaxBytesError
		if errors.As(err, &tooLarge) {
			writeJSON(w, http.StatusRequestEntityTooLarge, errorBody{
				Detail: fmt.Sprintf("File exceeds the %d MB limit", s.cfg.MaxUploadMB)})
			return
		}
		writeError(w, clientError(errs.ErrBadRequest, "expected a multipart upload with a 'file' part"))
		return
	}
	file, header, err := r.FormFile("file")
	if err != nil {
		writeError(w, clientError(errs.ErrBadRequest, "no 'file' part in the upload"))
		return
	}
	defer file.Close()

	data, err := io.ReadAll(file)
	if err != nil {
		writeError(w, clientError(errs.ErrBadRequest, "could not read the upload"))
		return
	}
	if int64(len(data)) > limit {
		writeJSON(w, http.StatusRequestEntityTooLarge,
			errorBody{Detail: fmt.Sprintf("File exceeds the %d MB limit", s.cfg.MaxUploadMB)})
		return
	}
	if len(data) == 0 {
		writeError(w, clientError(errs.ErrBadRequest, "Empty upload"))
		return
	}

	if repo.IsPDF(data) {
		// Called out separately because people WILL try it, and "unsupported image type"
		// does not tell them what to do about it.
		writeJSON(w, http.StatusUnsupportedMediaType, errorBody{
			Detail: "PDF is not supported — upload a JPEG, PNG, WEBP, BMP or TIFF image"})
		return
	}
	ext, mediaType, ok := repo.SniffImage(data)
	if !ok {
		// Sniffed from MAGIC BYTES, not the client's Content-Type, which is
		// attacker-controlled and wrong often enough to be useless.
		writeJSON(w, http.StatusUnsupportedMediaType,
			errorBody{Detail: "Unsupported file type — expected an image"})
		return
	}
	width, height, decoded := repo.DecodeDimensions(data)
	if !decoded {
		writeError(w, clientError(errs.ErrImageUnreadable, "The image could not be decoded — it may be corrupt"))
		return
	}

	filename := safeFilename(header.Filename)

	// **BYTES FIRST, ROW SECOND.** The record is what makes the document visible to the
	// worker, so writing it before the file leaves a window in which the drain loop can
	// claim a document whose original does not exist yet — reporting a perfectly good
	// upload as failed. See repo.ReserveID.
	id := repo.ReserveID(s.db)
	if _, err := repo.SaveOriginal(s.db, id, data, ext); err != nil {
		writeError(w, err)
		return
	}
	rec := model.NewDocument(id, filename, mediaType, int64(len(data)), ext)
	rec.OriginalW, rec.OriginalH = &width, &height
	rec.SearchText = strings.ToLower(filename)
	rec = repo.Create(s.db, rec)

	s.worker.NotifyNewWork()
	slog.Info("[API] queued document", "doc", rec.ID, "filename", filename, "bytes", len(data))

	out := row(rec)
	if pos, ok := repo.QueuePosition(s.db, rec.ID); ok {
		out["queue_position"] = pos
	} else {
		out["queue_position"] = nil
	}
	writeJSON(w, http.StatusAccepted, out)
}

// handleList serves one page of the document log.
func (s *Server) handleList(w http.ResponseWriter, r *http.Request, _ *Identity) {
	q := r.URL.Query()

	// The filter parameter is named `status` on the wire. Keeping that name is a client
	// dependency, not a preference.
	statusFilter := q.Get("status")
	if statusFilter != "" && !model.ValidStatuses[statusFilter] {
		writeError(w, clientError(errs.ErrBadRequest, "Invalid status"))
		return
	}
	sortDir := q.Get("sort_dir")
	if sortDir != "asc" && sortDir != "desc" {
		sortDir = "desc"
	}
	// Bounds copied from the reference's own declarations (service/api/documents.py:173-174):
	// page is ge=1 with NO upper bound, page_size is ge=1 le=100. Out of range is a 422, not a
	// clamp — see params.go.
	page, err := queryInt(q, "page", 1, 1, 0)
	if err != nil {
		writeError(w, err)
		return
	}
	pageSize, err := queryInt(q, "page_size", 20, 1, 100)
	if err != nil {
		writeError(w, err)
		return
	}

	rows, total := repo.GetAll(s.db, store.Query{
		Status:   statusFilter,
		DocType:  q.Get("doc_type"),
		Search:   q.Get("search"),
		DateFrom: q.Get("date_from"),
		DateTo:   q.Get("date_to"),
		Page:     page,
		PageSize: pageSize,
		SortBy:   q.Get("sort_by"),
		SortDir:  sortDir,
	})

	items := make([]map[string]any, 0, len(rows))
	for _, rec := range rows {
		items = append(items, row(rec))
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"items":     items,
		"total":     total,
		"page":      page,
		"page_size": pageSize,
		"stats":     repo.Stats(s.db),
	})
}

// clampInt is gone: silent clamping answered 200 to requests the reference rejects with 422.
// Bounded query integers go through queryInt in params.go, which reproduces the reference's
// pydantic-shaped rejection.

func (s *Server) handleGetDocument(w http.ResponseWriter, r *http.Request, _ *Identity, id int) {
	rec := repo.GetByID(s.db, id)
	if rec == nil {
		writeError(w, clientError(errs.ErrNotFound, "Document not found"))
		return
	}
	writeJSON(w, http.StatusOK, detail(rec))
}

// handleProgress returns live progress, a queue position, or a terminal state.
//
// **200 with a JSON null when there is nothing to report — never 404.** The polling client
// would otherwise raise an error toast every two seconds for a document that finished
// perfectly well.
func (s *Server) handleProgress(w http.ResponseWriter, r *http.Request, _ *Identity, id int) {
	rec := repo.GetByID(s.db, id)
	if rec == nil {
		writeError(w, clientError(errs.ErrNotFound, "Document not found"))
		return
	}

	if live := s.worker.DocumentProgress(id); live != nil {
		writeJSON(w, http.StatusOK, live)
		return
	}

	switch rec.Status {
	case model.StatusQueued:
		position := 0
		if p, ok := repo.QueuePosition(s.db, id); ok {
			position = p
		}
		pos := position
		writeJSON(w, http.StatusOK, map[string]any{
			"step":  "queued",
			"label": fmt.Sprintf("Queued (#%d)", position+1),
			"pct":   0,
			// The estimate is "everything ahead of me at the current average", which is
			// honest about being a guess and tracks reality because the average is an EMA
			// of real completions.
			"eta_sec":        round1(float64(position) * s.worker.AverageDurationSec()),
			"queue_position": &pos,
		})
	case model.StatusDone, model.StatusFailed:
		pct := 0
		if rec.Status == model.StatusDone {
			pct = 100
		}
		writeJSON(w, http.StatusOK, map[string]any{
			"step":           rec.Status,
			"label":          strings.ToUpper(rec.Status[:1]) + rec.Status[1:],
			"pct":            pct,
			"eta_sec":        nil,
			"queue_position": nil,
		})
	default:
		// A JSON null body, deliberately. See the function note.
		writeJSON(w, http.StatusOK, nil)
	}
}

func round1(v float64) float64 {
	return float64(int(v*10+0.5)) / 10
}

// handleImage serves an artifact.
//
// `no-cache` means REVALIDATE, not "do not store": http.ServeFile still sends ETag and
// Last-Modified, so a repeat request costs a 304 with no body. `max-age` would be wrong here
// — Reprocess overwrites canvas.png and thumb.jpg at the SAME URL, so the browser would keep
// showing the previous recognition's image while the field table beside it was already new.
func (s *Server) handleImage(w http.ResponseWriter, r *http.Request, _ *Identity,
	id int, kind string) {

	switch kind {
	case "original", "canvas", "thumb":
	default:
		writeError(w, clientError(errs.ErrNotFound, "Unknown image kind"))
		return
	}
	path, mediaType, ok := repo.OpenArtifact(s.db, id, kind)
	if !ok {
		writeError(w, clientError(errs.ErrNotFound, "Image not available"))
		return
	}
	w.Header().Set("Content-Type", mediaType)
	w.Header().Set("Cache-Control", "private, no-cache")
	http.ServeFile(w, r, path)
}

func (s *Server) handleReprocess(w http.ResponseWriter, r *http.Request, _ *Identity, id int) {
	rec := repo.GetByID(s.db, id)
	if rec == nil {
		writeError(w, clientError(errs.ErrNotFound, "Document not found"))
		return
	}
	if repo.ActiveStatuses[rec.Status] {
		writeError(w, clientError(errs.ErrConflict, "Document is already %s", rec.Status))
		return
	}
	rec = repo.Requeue(s.db, rec)
	s.worker.NotifyNewWork()
	writeJSON(w, http.StatusOK, row(rec))
}

// handleDelete returns 204 and an EMPTY BODY.
func (s *Server) handleDelete(w http.ResponseWriter, r *http.Request, _ *Identity, id int) {
	rec := repo.GetByID(s.db, id)
	if rec == nil {
		writeError(w, clientError(errs.ErrNotFound, "Document not found"))
		return
	}
	repo.Delete(s.db, rec)
	writeNoContent(w)
}

// handlePurge clears the scratch store. SESSION ONLY — not something an integration does.
func (s *Server) handlePurge(w http.ResponseWriter, r *http.Request, _ *Identity) {
	removed := 0
	for _, rec := range s.db.AllRecords() {
		// An in-flight job is left alone: deleting its record while the worker holds it
		// produces a "document vanished" failure that looks like a bug.
		if rec.Status == model.StatusProcessing {
			continue
		}
		repo.Delete(s.db, rec)
		removed++
	}
	slog.Info("[API] purged documents", "count", removed)
	writeJSON(w, http.StatusOK, map[string]any{"deleted": removed})
}
